### DISCUSSION TITLE
[Released] Integration Request: NewRelic

### DISCUSSION THREAD

**Discussion by stephenhmarsh** [vercel/community#1067] [Idea]
[The Logging page](https://vercel.com/guides/how-do-i-store-logs-on-vercel) said I should open an "Integration Request" so here it is:

We would love to have NewRelic support both as an APM and also a logging integration.  See also: https://github.com/newrelic/newrelic-node-nextjs/issues/94

**Comment by amyegan** [comment]
Thanks for reaching out @stephenhmarsh! A [New Relic integration is in development](https://twitter.com/vercel/status/1583231795750789121?s=20&t=ulXA-1xLcq4PiF7KebyLLw), but not quite ready yet.

**Reply by amyegan** [reply]
It's available now! [vercel.com/integrations/newrelic](https://vercel.com/integrations/newrelic)

**Reply by andenacitelli** [reply]
The integration works great for streaming logs over to NewRelic. Thanks for the hard work. Any update on getting the APM / Tracing part working? I've looked basically across the entire internet and haven't yet found someone that successfully got APM for Serverless Next.js functions (i.e. handlers in the `/api` directory) working. Thank you!

**Reply by amyegan** [reply]
Glad to hear it's working well for you! 😄  I don't have a specific ETA available yet, but our engineers are working with New Relic to enable APM support.

**Reply by gnohj** [reply]
@amyegan any progress on the APM support?

**Reply by amyegan** [reply]
APM traces can be monitored through [OpenTelemetry](https://vercel.com/docs/concepts/observability/otel-overview/quickstart) with the [New Relic integration](https://vercel.com/integrations/newrelic). New Relic has a quickstart guide with more info about available features: https://newrelic.com/instant-observability/vercel

**Reply by gnohj** [reply]
@amyegan does this feature support other frameworks (Remix.js) or is it exclusive for only Next.js?

**Reply by amyegan** [reply]
The integration supports just about anything you would deploy with Vercel. OpenTelemetry support is still in beta and may not work perfectly with all frameworks yet. If you try it and run into a problem, please let us know 🙏

**Reply by gnohj** [reply]
@amyegan I am following https://vercel.com/docs/concepts/observability/otel-overview/quickstart. Not seeing any APM traces, let alone any APM Service created in our NewRelic Account. I am also confused on how this would work with loaders/actions in Remix. Since we have to do an explicit return for the payload, ex:

```
export const loader = () => {
    const span = tracer.startSpan('root', undefined, context.active());
    .....
    span.end(); <-- has to end before payload gets returned
    return json({ data });
}
```

In Next.js, there is no explicit return since its using the response obj, ex:

```
export default async function handler(_request, response) {
   const span = tracer.startSpan('handler', undefined , context.active());
   const ms = Math.floor(Math.random() * 1000);
   span.setAttribute('sleep', ms);
   await new Promise((resolve) =>setTimeout(resolve, ms));
   response.status(200).json({ greetings: `Hi there ${ms}!`});
   span.end();
}
```
